### PR TITLE
Allow host to subscribe to logs from enclave, and send back logs to subscribers

### PR DESCRIPTION
**Pull request by tudor-malene**
### Why is this change needed?

We need to stream events based on the rules described in the events design doc

### What changes were made as part of this PR:

To keep this PR smaller, many changes are skipped in this PR (search "TODO - #453" for a full list). These include:

- No encryption of subscription results
- No filtering of subscription results (if the host creates any subscriptions, it will get _all_ logs)
- No API to route logs from the host to the client
- No checking that a host has the right to create a specific subscription

It does introduce the following changes:

- APIs to allow the host to (un)subscribe for logs from the enclave
- Management of the subscriptions
- Inclusion of the logs in the block submission responses for the host (iff they are subscribed)

To test this is working (tests won't be possible until the client <-> host APIs are introduced), you can:

* Create a subscription by adding this code to line 221 of `go/host/node/node.go`:

```
	println("jjj attempting to subscribe")
	id, err := uuid.NewUUID()
	if err != nil {
		panic(err)
	}
	subscription := common.LogSubscription{
		Accounts: []*common.SubscriptionAccount{},
	}
	encodedSubscription, err := rlp.EncodeToBytes(subscription)
	if err != nil {
		panic(err)
	}
	err = a.enclaveClient.Subscribe(id, encodedSubscription)
	if err != nil {
		panic(err)
	}
	println("jjj subscribed")
```

* Check that the host is receiving logs by adding this code to line 582 of `go/host/node/node.go`:

```
	if len(result.SubscribedLogs) != 0 {
		for _, logs := range result.SubscribedLogs {
			println(fmt.Sprintf("jjj got logs: %v", logs))
		}
	}
```

You will then see the logs rolling in if you run the full network sim.

### :rotating_light: Definition of Done :rotating_light:
- [ ] Deployed into dev-testnet 
- [ ] Passes tests on dev-testnet
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated

### CODE REVIEW

File: go/enclave/rollupchain/rollup_chain.go [hunk 1]
```diff
@@ -159,11 +167,12 @@ func (rc *RollupChain) newBlockSubmissionResponse(bs *obscurocore.BlockState, ro
...
-		ProducedRollup: rollup,
-		IngestedBlock:  true,
-		FoundNewHead:   bs.FoundNewRollup,
-		RollupHead:     head,
+		BlockHeader:        headBlock.Header(),
+		ProducedRollup:     rollup,
+		IngestedBlock:      true,
+		FoundNewHead:       bs.FoundNewRollup,
+		RollupHead:         head,
+		SubscribedReceipts: receipts,
```
Comment (joeldudley) [inline go/enclave/rollupchain/rollup_chain.go:175]: @tudor-malene Here and elsewhere, shouldn't this specifically be logs, and not the entire receipt?

**Comment by joeldudley** [comment]
@tudor-malene I realised that in the design doc, we weren't specific about which event types we were handling. There are three types in the spec - `newPendingTransactions`, `newHeads` and `logs`. I assume we're only talking about logs in your design? Should I update the design doc accordingly?

File: go/enclave/rollupchain/rollup_chain.go (hunk 1 above)
Comment (tudor-malene) [reply go/enclave/rollupchain/rollup_chain.go:175]: yes, should be logs everywhere.
changing now.

**Comment by tudor-malene** [comment]
> @tudor-malene I realised that in the design doc, we weren't specific about which event types we were handling. There are three types in the spec - `newPendingTransactions`, `newHeads` and `logs`. I assume we're only talking about logs in your design? Should I update the design doc accordingly?

yes please. The design is strictly about logs.
But the work of including the plumbing for subscriptions, should probably cover all of these.

**Review by BedrockSquirrel — APPROVED** [review]
LGTM generally

I didn't see a todo about subscription manager on the host side but I guess that will need to exist as well? And I suspect host subscription manager will need to cancel subscriptions on the enclave (e.g. on disconnect), so might not be able to require signature for subscription cancel.

File: go/enclave/rollupchain/rollup_chain.go [hunk 2]
```diff
@@ -226,7 +250,27 @@ func (rc *RollupChain) updateState(b *types.Block) *obscurocore.BlockState {
...
 
-	return bs
+	var logs []*types.Log
+	for _, receipt := range receipts {
+		logs = append(logs, receipt.Logs...)
+	}
+
+	subscribedLogs := rc.subscriptionManager.FilterRelevantLogs(logs)
+
+	// TODO - #453 - Check this recursive logic works correctly (i.e. each rollup contains the logs of all its ancestors as well).
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/enclave/rollupchain/rollup_chain.go:260]: Yeah, they do. This was how Tudor laid it out, but worth challenging when he gets back.

**Pull request merged**